File: sections/28_dp_i_mixed_contest/problems/i_grouped_cargo/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    index = 0
    groups, capacity = data[index:index + 2]
    index += 2
    dp = [0] * (capacity + 1)

    for _ in range(groups):
        count = data[index]
        index += 1
        items = []
        for _ in range(count):
            weight, value = data[index:index + 2]
            index += 2
            items.append((weight, value))

        next_dp = dp.copy()
        for weight, value in items:
            for used in range(capacity - weight + 1):
                next_dp[used + weight] = max(
                    next_dp[used + weight], dp[used] + value)
        dp = next_dp

    print(max(dp))


if __name__ == "__main__":
    main()

File: sections/28_dp_i_mixed_contest/problems/i_grouped_cargo/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int groups, capacity;
    cin >> groups >> capacity;
    vector<long long> dp(capacity + 1, 0);

    for (int group = 0; group < groups; ++group) {
        int count;
        cin >> count;
        vector<pair<int, long long>> items(count);
        for (auto& [weight, value] : items)
            cin >> weight >> value;

        vector<long long> next = dp;
        for (auto [weight, value] : items)
            for (int used = 0; used + weight <= capacity; ++used)
                next[used + weight] = max(next[used + weight],
                                          dp[used] + value);
        dp.swap(next);
    }

    cout << *max_element(dp.begin(), dp.end()) << '\n';
}
